### PR TITLE
feat: added presets to canvas resize window 

### LINKED ISSUE #3181: Feature request: Presets in Canvas Resize dialog
Synfig has presets feature in its Preferences to set the default canvas size. It'd be helpful if similar presets are also made available in Canvas resize dialog.
image

Screenshot 2023-08-16 160844

### PR DESCRIPTION
**Pull request by Afterdie**
Сloses #3181
Didn't add  the options in glade file since the widget in ```dialog_setup.cpp``` follows a similar approach.

https://github.com/synfig/synfig/assets/5604544/cf838665-6f64-4eee-bb63-58e2bb985be0

image

### CODE REVIEW

File: synfig-studio/src/gui/dialogs/canvasresize.cpp [hunk 1]
```diff
@@ -259,6 +277,28 @@ void CanvasResize::on_height_changed()
 	refresh_wh_toggle_widgets();
 }
 
+void CanvasResize::on_size_template_changed(){
+	String selection(combo_box->get_active_text());
+	if(selection == "Custom Size"){
```
Comment (rodolforg) [inline synfig-studio/src/gui/dialogs/canvasresize.cpp:282]: It won't work with the translated string. You should set the combobox item id (with a non translatable string) and check it here.

**Review by rodolforg — CHANGES_REQUESTED** [review]

File: synfig-studio/src/gui/dialogs/canvasresize.cpp [hunk 2]
```diff
@@ -124,6 +125,21 @@ void CanvasResize::set_up_builder_widgets()
...
+	combo_box->prepend(_("720x576   DVD PAL"));
+	combo_box->prepend(_("720x480   DVD NTSC"));
+	combo_box->prepend(_("720x540   Web 720x"));
+	combo_box->prepend(_("720x405   Web 720x HD"));
+	combo_box->prepend(_("640x480   Web 640x"));
+	combo_box->prepend(_("640x360   Web 640x HD"));
+	combo_box->prepend(_("480x360   Web 480x"));
+	combo_box->prepend(_("480x270   Web 480x HD"));
+	combo_box->prepend(_("360x270   Web 360x"));
+	combo_box->prepend(_("360x203   Web 360x HD"));
```
[the remark this answers is not in the document]
Comment (Afterdie) [reply synfig-studio/src/gui/dialogs/canvasresize.cpp:142]: The code in ```dialog_setup.cpp``` had a similar approach so i thought this was the convention. Totally agree about the id part ill work on it.

[the remark this answers is not in the document]
Comment (Afterdie) [reply synfig-studio/src/gui/dialogs/canvasresize.cpp:142]: https://github.com/synfig/synfig/blob/7040b7f4452afebe131a4425bafb5171f7f6112f/synfig-studio/src/gui/dialogs/dialog_setup.cpp#L351 for reference

File: synfig-studio/src/gui/dialogs/canvasresize.cpp [hunk 3]
```diff
@@ -259,6 +262,28 @@ void CanvasResize::on_height_changed()
 	refresh_wh_toggle_widgets();
 }
 
+void CanvasResize::on_size_template_changed(){
+	String selection(combo_box->get_active_id());
+	if(selection == "custom"){
```
Comment (rodolforg) [inline synfig-studio/src/gui/dialogs/canvasresize.cpp:267]: prefer 
```c++ 
if (selection == "custom") {
``` 
 (pay attention at blank spaces)

File: synfig-studio/src/gui/dialogs/canvasresize.cpp [hunk 4]
```diff
@@ -259,6 +262,28 @@ void CanvasResize::on_height_changed()
...
+	String y_size(selection.substr(locx+1,strSize-1));
+	int x=stoi(x_size);
+	int y=stoi(y_size);
+	toggle_ratio_wh->set_active(false);
+	width->set_value(x);
+	height->set_value(y);
+	width->set_sensitive(false);
+	height->set_sensitive(false);
+	toggle_ratio_wh->set_active(true);
+	return;
```
Comment (rodolforg) [inline synfig-studio/src/gui/dialogs/canvasresize.cpp:284]: you can remove this `return;` ;)

**Review by rodolforg — CHANGES_REQUESTED** [review]

File: synfig-studio/src/gui/dialogs/canvasresize.cpp [hunk 5]
```diff
@@ -259,6 +262,27 @@ void CanvasResize::on_height_changed()
 	refresh_wh_toggle_widgets();
 }
 
+void CanvasResize::on_size_template_changed(){
+	String selection(combo_box->get_active_id());
+	if(selection == "custom") {
```
Comment (rodolforg) [inline synfig-studio/src/gui/dialogs/canvasresize.cpp:267]: prefer
```c++
if (selection == "custom") {
```
(pay attention at blank spaces)

File: synfig-studio/src/gui/dialogs/canvasresize.cpp [hunk 6]
```diff
@@ -259,6 +262,27 @@ void CanvasResize::on_height_changed()
...
+		width->set_sensitive(true);
+		height->set_sensitive(true);
+		return;
+	}
+	String::size_type locx=selection.find_first_of('x');
```
Comment (rodolforg) [inline synfig-studio/src/gui/dialogs/canvasresize.cpp:272]: use spaces surrounding equal sign `=`

File: synfig-studio/src/gui/dialogs/canvasresize.cpp [hunk 7]
```diff
@@ -259,6 +262,27 @@ void CanvasResize::on_height_changed()
...
+	String::size_type strSize=selection.size();
+	String x_size(selection.substr(0,locx));
```
Comment (rodolforg) [inline synfig-studio/src/gui/dialogs/canvasresize.cpp:274]: Just for safety, check if `locx` isn't `std::string::npos` (i.e., if character 'x' was found).

File: synfig-studio/src/gui/dialogs/canvasresize.cpp [hunk 8]
```diff
@@ -259,6 +262,27 @@ void CanvasResize::on_height_changed()
...
+	String y_size(selection.substr(locx+1,strSize-1));
```
Comment (rodolforg) [inline synfig-studio/src/gui/dialogs/canvasresize.cpp:275]: It gives us problem if `locx` is `std::string::npos`

File: synfig-studio/src/gui/dialogs/canvasresize.cpp [hunk 9]
```diff
@@ -259,6 +262,27 @@ void CanvasResize::on_height_changed()
...
+	int x=stoi(x_size);
+	int y=stoi(y_size);
```
Comment (rodolforg) [inline synfig-studio/src/gui/dialogs/canvasresize.cpp:277]: `stoi()` [raises exception](https://en.cppreference.com/w/cpp/string/basic_string/stol) if string can't be converted. It is a controlled environment (internal ComboBox item id), but...

**Review by rodolforg — CHANGES_REQUESTED** [review]

File: synfig-studio/src/gui/dialogs/canvasresize.cpp (hunk 9 above)
Comment (Afterdie) [reply synfig-studio/src/gui/dialogs/canvasresize.cpp:277]: should i use long instead?

Comment (rodolforg) [reply synfig-studio/src/gui/dialogs/canvasresize.cpp:277]: No, you don't need to. Well, the exception won't be thrown if the combobox item ids are well formed, so it is just  preciousness of mine. Let it be as it is.

**Review by rodolforg — APPROVED** [review]

**Comment by ice0** [comment]
Merged. Thank you!

**Pull request merged**